### DISCUSSION TITLE
How to get typescript typings for static methods when changing database connections

### DISCUSSION THREAD

**Discussion by kkerti** [typegoose/typegoose#797] [Q&A] [answered]
I have the following database changing class wrapper:

```ts
import { getModelForClass } from "@typegoose/typegoose";
import { AnyParamConstructor } from "@typegoose/typegoose/lib/types";
import mongoose from "mongoose";

function getTheConnection(dbName: string) {
    const index = mongoose.connections.findIndex((connection) => connection.name == dbName)
    return mongoose.connections[index].useDb(dbName)
}

export function DBModel<T>(databaseName: string, modelClass: AnyParamConstructor<T>) {
    const model = getModelForClass(modelClass)
    const db = getTheConnection(databaseName)
    const DbModel = db.model(model.modelName, model.schema);
    return DbModel
}
```

Then I have some static functions in a class, like this:

```ts
@ModelOptions({ schemaOptions: { timestamps: true, collection: 'inventory' } })
export class Inventory extends BaseInventory {
	public static async listItemsOnStock(this: ReturnModelType<typeof Inventory>, itemName: string) {
		return await this.find({ done: true, fulfillmentId: null, deleted: { $not: { $eq: true } }, name: { $in: itemName.split(',') } })
	}
}
```

When I use the DBModel wrapper to define the database I want to use, I get type errors for the static method:

<img width="1139" alt="image" src="https://user-images.githubusercontent.com/47832952/224506919-86ac12be-2eb1-42b2-b8d1-6a070c4f6a8b.png">

---
How can I get proper typing for the static methods?

**Comment by hasezoey** [comment, marked as answer]
>     const DbModel = db.model(model.modelName, model.schema);

i dont think that will give your the proper type, you could also just use `getModelForClass(Class, { existingConnection: db });`
(if you want to be able to use the same class for multiple connections, then consider [disabling caching](https://typegoose.github.io/typegoose/docs/guides/advanced/models-with-same-name#disable-caching))

if you still want to use the current way you have, then maybe cast your return like `return yourModel as ReturnModelType<modelClass>` (maybe requires to use `typeof modelClass`, didnt test)

**Reply by kkerti** [reply]
I have followed your suggestion and redesigned my pattern. Now I have a simple fastify plugin, where all my needed models are initialized with a database connection passed to the `existingConnection` property. Thank you for the suggestion!
